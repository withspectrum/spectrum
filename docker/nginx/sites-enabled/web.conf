server {
    listen 80;

    server_name www.spectrum.localhost spectrum.localhost;

    location / {
        resolver 127.0.0.11 ipv6=off;
        set $service spectrum:3000;
        proxy_pass http://$service;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection 'upgrade';
        proxy_set_header Host $host;
        proxy_cache_bypass $http_upgrade;
    }
}

