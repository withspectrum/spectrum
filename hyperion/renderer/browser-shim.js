// @flow
// Shim some browser stuff we use in the client for server-side rendering
// NOTE(@mxstbr): We should be getting rid of this over time
global.window = {
  location: {
    protocol: 'https:',
    host: 'spectrum.chat',
    hash: '',
  },
};
global.localStorage = {
  getItem: () => null,
  setItem: () => {},
  removeItem: () => {},
};
global.navigator = {
  userAgent: '',
};
