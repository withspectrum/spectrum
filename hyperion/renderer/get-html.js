// @flow
import fs from 'fs';
import path from 'path';
import serialize from 'serialize-javascript';

const getIndex = () => {
  return (
    fs
      .readFileSync(path.resolve(__dirname, '..', '..', 'build', 'index.html'))
      .toString()
      .replace(
        '<script type="text/javascript" src="/./static/js/bootstrap.js"></script>',
        ''
      )
      // This automatically gets injected without the defer tag in development by create-react-app, gotta get rid of it for dev to work!
      .replace(/(src="\/static\/js\/bundle\.js")/g, ' defer="defer" $1')
  );
};

let html = getIndex();

type Arguments = {
  styleTags: string,
  metaTags: string,
  state: Object,
  content: string,
  scriptTags: string,
};

export const createScriptTag = ({ src }: { src: string }) =>
  `<script defer="defer" src="${src}"></script>`;

const sentry = `<script defer="defer" src="https://cdn.ravenjs.com/3.14.0/raven.min.js" crossorigin="anonymous"></script><script defer="defer" src="/install-raven.js"></script>`;

const polyfill = createScriptTag({
  src: 'https://cdn.polyfill.io/v2/polyfill.min.js',
});

export const getHTML = ({
  styleTags,
  metaTags,
  state,
  content,
  scriptTags,
}: Arguments) => {
  if (process.env.NODE_ENV === 'development') html = getIndex();
  return (
    html
      // Inject the state and the content instead of <div id="root">
      .replace(
        '<div id="root"></div>',
        `<script>window.__SERVER_STATE__=${serialize(
          state
        )}</script><div id="root">${content}</div><script type="text/javascript" src="/./static/js/bootstrap.js"></script>${sentry}${polyfill}${scriptTags}`
      )
      // Inject the meta tags at the start of the <head>
      .replace('<head>', `<head>${metaTags}`)
      // Inject the style tags at the end of the <head>
      .replace('</head>', `${styleTags}</head>`)
  );
};
