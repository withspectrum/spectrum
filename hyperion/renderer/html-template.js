// @flow
import fs from 'fs';
import path from 'path';
import { html } from 'common-tags';
import serialize from 'serialize-javascript';

// Match main.asdf123.js in production mode or bundle.js in dev mode
const mainBundleRegex = /(main|bundle)\.(?:.*\.)?js$/;

let bundles;
try {
  bundles = fs.readdirSync(path.join(__dirname, '../../build/static/js'));
} catch (err) {
  throw new Error(
    'It looks like you didn\'t run "yarn run dev:web" or "yarn run build:web" before starting hyperion. Please wait until either of them completes before starting hyperion.'
  );
}

// Get the main bundle filename
const mainBundle = bundles.find(bundle => mainBundleRegex.test(bundle));
if (!mainBundle) {
  throw new Error(
    'It looks like you didn\'t run "yarn run dev:web" or "yarn run build:web" before starting hyperion. Please wait until either of them completes before starting hyperion.'
  );
}

export const createScriptTag = ({ src }: { src: string }) =>
  `<script defer="defer" src="${src}"></script>`;

export const getHeader = ({ metaTags }: { metaTags: string }) => {
  return html`
      <!DOCTYPE html>
      <html lang="en">
        <head>
          <meta charset="utf-8">
          <meta name="viewport" content="width=device-width,initial-scale=1">
          <link rel="mask-icon" href="/img/pinned-tab.svg" color="#171A21">
          <meta name="theme-color" content="#171A21">
          <link rel="manifest" href="/manifest.json">
          <meta name="og:type" content="website">
          <meta name="og:site_name" content="Spectrum.chat">
          <meta name="twitter:card" content="summary">
          <meta name="twitter:site" content="@withspectrum">
          <meta name="twitter:image:alt" content="Where communities are built">
          <link rel="apple-touch-icon-precomposed" sizes="57x57" href="/img/apple-icon-57x57-precomposed.png" />
          <link rel="apple-touch-icon-precomposed" sizes="72x72" href="/img/apple-icon-72x72-precomposed.png" />
          <link rel="apple-touch-icon-precomposed" sizes="114x114" href="/img/apple-icon-114x114-precomposed.png" />
          <link rel="apple-touch-icon-precomposed" sizes="144x144" href="/img/apple-icon-144x144-precomposed.png" />
          ${metaTags}
          <script>
              !function(e,a,t,n,g,c,o){e.GoogleAnalyticsObject=g,e.ga=e.ga||function(){(e.ga.q=e.ga.q||[]).push(arguments)},e.ga.l=1*new Date,c=a.createElement(t),o=a.getElementsByTagName(t)[0],c.defer=1,c.src="https://www.google-analytics.com/analytics.js",o.parentNode.insertBefore(c,o)}(window,document,"script",0,"ga"),ga("create","UA-92673909-1","auto"),ga("send","pageview"),ga('set', 'anonymizeIp', true)
          </script>
          <script type="text/javascript">
            (function(e,t){var n=e.amplitude||{_q:[],_iq:{}};var r=t.createElement("script")
            ;r.type="text/javascript";r.async=true
            ;r.src="https://cdn.amplitude.com/libs/amplitude-4.2.1-min.gz.js"
            ;r.onload=function(){if(e.amplitude.runQueuedFunctions){
            e.amplitude.runQueuedFunctions()}else{
            console.error("[Amplitude] Error: could not load SDK")}}
            ;var i=t.getElementsByTagName("script")[0];i.parentNode.insertBefore(r,i)
            ;function s(e,t){e.prototype[t]=function(){
            this._q.push([t].concat(Array.prototype.slice.call(arguments,0)));return this}}
            var o=function(){this._q=[];return this}
            ;var a=["add","append","clearAll","prepend","set","setOnce","unset"]
            ;for(var u=0;u<a.length;u++){s(o,a[u])}n.Identify=o;var c=function(){this._q=[]
            ;return this}
            ;var l=["setProductId","setQuantity","setPrice","setRevenueType","setEventProperties"]
            ;for(var p=0;p<l.length;p++){s(c,l[p])}n.Revenue=c
            ;var d=["init","logEvent","logRevenue","setUserId","setUserProperties","setOptOut","setVersionName","setDomain","setDeviceId","setGlobalUserProperties","identify","clearUserProperties","setGroup","logRevenueV2","regenerateDeviceId","logEventWithTimestamp","logEventWithGroups","setSessionId","resetSessionId"]
            ;function v(e){function t(t){e[t]=function(){
            e._q.push([t].concat(Array.prototype.slice.call(arguments,0)))}}
            for(var n=0;n<d.length;n++){t(d[n])}}v(n);n.getInstance=function(e){
            e=(!e||e.length===0?"$default_instance":e).toLowerCase()
            ;if(!n._iq.hasOwnProperty(e)){n._iq[e]={_q:[]};v(n._iq[e])}return n._iq[e]}
            ;e.amplitude=n})(window,document);
          </script>
        </head>
        <body>
          <div id="root">`;
};

export const getFooter = ({
  state,
  data,
  bundles,
}: {
  state: Object,
  data: Object,
  bundles: Array<string>,
}) => {
  return html`</div>
      <script defer="defer" src="https://cdn.ravenjs.com/3.14.0/raven.min.js" crossorigin="anonymous"></script>
      <script defer="defer" src="/install-raven.js"></script>
      <script>window.__SERVER_STATE__=${serialize(state)}</script>
      <script>window.__DATA__=${serialize(data)}</script>
      <script defer="defer" type="text/javascript" src="https://cdn.polyfill.io/v2/polyfill.min.js?features=default,Array.prototype.find,Symbol.iterator"></script>
      <script type="text/javascript" src="/static/js/bootstrap.js"></script>
      ${bundles.map(src => createScriptTag({ src }))}
      ${createScriptTag({ src: `/static/js/${mainBundle}` })}
    </body>
    </html>
  `;
};
