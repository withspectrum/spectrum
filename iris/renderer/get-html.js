// @flow
import fs from 'fs';
import path from 'path';
import serialize from 'serialize-javascript';

const html = fs
  .readFileSync(path.resolve(__dirname, '..', '..', 'build', 'index.html'))
  .toString();

type Arguments = {
  styleTags: string,
  metaTags: string,
  state: Object,
  content: string,
};

const sentry = `<script src="https://cdn.ravenjs.com/3.14.0/raven.min.js" crossorigin="anonymous"></script><script>Raven.config('https://3bd8523edd5d43d7998f9b85562d6924@sentry.io/154812', { whitelistUrls: [/spectrum.chat/, /www.spectrum.chat/] }).install();</script>`;

export const getHTML = ({ styleTags, metaTags, state, content }: Arguments) => {
  return (
    html
      // Inject the state and the content instead of <div id="root">
      .replace(
        '<div id="root"></div>',
        `<script>window.__SERVER_STATE__=${serialize(
          state
        )}</script><div id="root">${content}</div>`
      )
      // Inject the meta tags at the start of the <head>
      .replace('<head>', `<head>${metaTags}`)
      // Inject the style tags at the end of the <head>
      .replace('</head>', `${styleTags}${sentry}</head>`)
  );
};
