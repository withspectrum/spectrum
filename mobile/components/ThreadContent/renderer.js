// @flow
import React from 'react';
import { View } from 'react-native';
import styled from 'styled-components/native';
import redraft from 'redraft';
import Anchor from '../Anchor';
import Text from '../Text';
import Codeblock from '../Codeblock';
import IFrame from '../IFrame';

const renderer = {
  inline: {
    BOLD: (children, { key }) => (
      <Text bold key={`bold-${key}`}>
        {children}
      </Text>
    ),
    ITALIC: (children, { key }) => (
      <Text italic key={`italic-${key}`}>
        {children}
      </Text>
    ),
    UNDERLINE: (children, { key }) => (
      <Text underline key={`underline-${key}`}>
        {children}
      </Text>
    ),
    CODE: (children, { key }) => (
      <Codeblock key={`codeblock-${key}`}>{children}</Codeblock>
    ),
  },
  entities: {
    // key is the entity key value from raw
    LINK: (children, data, { key }) => (
      <Anchor key={key} href={data.url}>
        {children}
      </Anchor>
    ),
    embed: (children, { src }, { key }) => {
      return <IFrame key={key} src={src} />;
    },
  },
  blocks: {
    fallback: (children, { keys }) => (
      <View key={keys.join('|')}>{children}</View>
    ),
    unstyled: (children, { keys }) =>
      children.map((child, index) => (
        <Text type="body" key={keys[index] || index}>
          {child}
        </Text>
      )),
    // Note: Headings are offset by one because we always assume the title
    // of the thread to be level 1, so a level 1 heading inside the thread
    // body has to be level 2
    'header-one': (children, { keys }) =>
      children.map((child, index) => (
        <Text type="title2" key={keys[index] || index}>
          {child}
        </Text>
      )),
    'header-two': (children, { keys }) =>
      children.map((child, index) => (
        <Text type="title3" key={keys[index] || index}>
          {child}
        </Text>
      )),
    // blockquote: (children, { keys }) =>
    'unordered-list-item': (children, { depth, keys }) => {
      return children.map((item, index) => (
        <Text key={keys[index] || index} type="body">
          {'\u2022'} {item}
        </Text>
      ));
    },
    'ordered-list-item': (children, { depth, keys }) => {
      return children.map((item, index) => (
        <Text key={keys[index] || index} type="body">
          {index}. {item}
        </Text>
      ));
    },
    'code-block': (children, { keys }) => (
      <Codeblock key={keys.join('|')}>{children}</Codeblock>
    ),
  },
};

export default (rawContentState: Object) =>
  redraft(rawContentState, renderer, {
    blockFallback: 'fallback',
  });
