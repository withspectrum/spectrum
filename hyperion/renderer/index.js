// Server-side renderer for our React code
import fs from 'fs';
const debug = require('debug')('hyperion:renderer');
import React from 'react';
import ReactDOM from 'react-dom/server';
import { ServerStyleSheet } from 'styled-components';
import { ApolloProvider, renderToStringWithData } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { createHttpLink } from 'apollo-link-http';
import {
  InMemoryCache,
  IntrospectionFragmentMatcher,
} from 'apollo-cache-inmemory';
import { StaticRouter } from 'react-router';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import Helmet from 'react-helmet';
import * as graphql from 'graphql';
import Loadable from 'react-loadable';
import { getBundles } from 'react-loadable/webpack';
import Raven from 'shared/raven';
import introspectionQueryResultData from 'src/api/schema.json';
import stats from '../../build/react-loadable.json';

import getSharedApolloClientOptions from 'shared/graphql/apollo-client-options';
import { getHTML, createScriptTag } from './get-html';

// Browser shim has to come before any client imports
import './browser-shim';
const Routes = require('../../src/routes').default;
import { initStore } from '../../src/store';

const IN_MAINTENANCE_MODE =
  process.env.REACT_APP_MAINTENANCE_MODE === 'enabled';
const IS_PROD = process.env.NODE_ENV === 'production';
const FORCE_DEV = process.env.FORCE_DEV;

if (!IS_PROD || FORCE_DEV) console.log('Querying API at localhost:3001/api');

const renderer = (req, res) => {
  debug(`server-side render ${req.url}`);
  debug(`querying API at https://${req.hostname}/api`);
  // HTTP Link for queries and mutations including file uploads
  const httpLink = createHttpLink({
    uri:
      IS_PROD && !FORCE_DEV
        ? `https://${req.hostname}/api`
        : 'http://localhost:3001/api',
    credentials: 'include',
    headers: {
      cookie: req.headers.cookie,
    },
  });

  const cache = new InMemoryCache({
    fragmentMatcher: new IntrospectionFragmentMatcher({
      introspectionQueryResultData,
    }),
    ...getSharedApolloClientOptions(),
  });

  // Create an Apollo Client with a local network interface
  const client = new ApolloClient({
    ssrMode: true,
    link: httpLink,
    cache,
  });
  // Define the initial redux state
  const initialReduxState = {
    users: {
      currentUser: req.user,
    },
  };
  // Create the Redux store
  const store = initStore(initialReduxState);
  let modules = [];
  const report = moduleName => {
    debug(`codesplitted module ${moduleName} used`);
    modules.push(moduleName);
  };
  const context = {};
  // The client-side app will instead use <BrowserRouter>
  const frontend = (
    <Loadable.Capture report={report}>
      <ApolloProvider client={client}>
        <Provider store={store}>
          <StaticRouter location={req.url} context={context}>
            <Routes maintenanceMode={IN_MAINTENANCE_MODE} />
          </StaticRouter>
        </Provider>
      </ApolloProvider>
    </Loadable.Capture>
  );
  // Initialise the styled-components stylesheet and wrap the app with it
  const sheet = new ServerStyleSheet();
  debug(`render frontend`);
  renderToStringWithData(sheet.collectStyles(frontend))
    .then(content => {
      if (context.url) {
        debug('found redirect on frontend, redirecting');
        // Somewhere a `<Redirect>` was rendered, so let's redirect server-side
        res.redirect(301, context.url);
        return;
      }
      // Get the resulting data
      const state = store.getState();
      const data = client.extract();
      const helmet = Helmet.renderStatic();
      if (IN_MAINTENANCE_MODE) {
        debug('maintainance mode enabled, sending 503');
        res.status(503);
        res.set('Retry-After', 3600);
      } else {
        res.status(200);
      }
      const bundles = getBundles(stats, modules)
        // Create <script defer> tags from bundle objects
        .map(bundle =>
          createScriptTag({ src: `/${bundle.file.replace(/\.map$/, '')}` })
        )
        // Make sure only unique bundles are included
        .filter((value, index, self) => self.indexOf(value) === index);
      debug('compile and send html');
      const scriptTags = [...bundles].join('\n');
      debug(`script tags: ${scriptTags}`);
      // Compile the HTML and send it down
      res.send(
        getHTML({
          content,
          state,
          data,
          styleTags: sheet.getStyleTags(),
          metaTags:
            helmet.title.toString() +
            helmet.meta.toString() +
            helmet.link.toString(),
          scriptTags,
        })
      );
      res.end();
    })
    .catch(err => {
      console.error(err);
      Raven.captureException(err);
      res.status(500);
      res.send('Oops, something went wrong. Please try again!');
    });
};

export default renderer;
