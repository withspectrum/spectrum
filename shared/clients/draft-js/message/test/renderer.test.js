// @flow
import { messageRenderer } from '../renderer.web';
import mentionsDecorator from '../../mentions-decorator/index.web';
import linksDecorator from '../../links-decorator/index.web';

describe('messageRenderer', () => {
  it('should render certain blocks', () => {
    expect(Object.keys(messageRenderer.blocks)).toMatchSnapshot();
  });

  it('should have decorators', () => {
    expect(messageRenderer.decorators).toContain(mentionsDecorator);
    expect(messageRenderer.decorators).toContain(linksDecorator);
  });
});
