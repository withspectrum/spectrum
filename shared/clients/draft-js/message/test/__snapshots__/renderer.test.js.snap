// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`messageRenderer should render certain blocks 1`] = `
Array [
  "unstyled",
  "code-block",
]
`;
